source/log_writer.c

```diff
@@ -0,0 +1,214 @@
+    struct aws_file_writer *impl = (struct aws_file_writer *) writer->impl;
+
+    size_t length = output->len;
+    if (fwrite(output->bytes, 1, length, impl->log_file) < length) {
+        return aws_io_translate_and_raise_file_write_error(errno);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static int s_aws_file_writer_cleanup_fn(struct aws_log_writer *writer) {
+    struct aws_file_writer *impl = (struct aws_file_writer *) writer->impl;
+
+    int result = (*impl->vtable->close_file_fn)(impl);
+
+    if (impl->base_file_name != NULL) {
+        aws_mem_release(writer->allocator, impl->base_file_name);
+    }
+
+    aws_mem_release(writer->allocator, impl);
+
+    return result;
+}
+
+static struct aws_log_writer_vtable s_aws_file_writer_vtable = {
+    .write_fn = s_aws_file_writer_write_fn,
+    .cleanup_fn = s_aws_file_writer_cleanup_fn
+};
+
+/*
+ * Shared internal init implementation
+ */
+static int s_aws_file_writer_init_internal(
+    struct aws_log_writer *writer,
+    struct aws_allocator *allocator,
+    const char *file_name,
+    struct aws_file_writer_vtable *vtable) {
+
+    /* Allocate and initialize the file writer */
+    struct aws_file_writer *impl = (struct aws_file_writer *)aws_mem_acquire(allocator, sizeof(struct aws_file_writer));
+    if (impl == NULL) {
+        return aws_raise_error(AWS_ERROR_OOM);
+    }
+
+    impl->vtable = vtable;
+    impl->log_file = NULL;
+    impl->base_file_name = NULL;
+
+    /* copy the file name, if necessary */
+    if (file_name != NULL) {
+        impl->base_file_name = aws_string_new_from_c_str(allocator, file_name);
+        if (impl->base_file_name == NULL) {
+            aws_mem_release(allocator, impl);
+            return AWS_OP_ERR;
+        }
+    }
+
+    /* attempt to open the file */
+    if ((*vtable->open_file_fn)(impl)) {
+        if (impl->base_file_name != NULL) {
+            aws_mem_release(allocator, impl->base_file_name);
+        }
+        aws_mem_release(allocator, impl);
+        return AWS_OP_ERR;
+    }
+
+    writer->vtable = &s_aws_file_writer_vtable;
+    writer->allocator = allocator;
+    writer->impl = impl;
+
+    return AWS_OP_SUCCESS;
+}
+
+/*
+ * Public initialization interface
+ */
+int aws_stdout_log_writer_init(struct aws_log_writer *writer, struct aws_allocator *allocator) {
+    return s_aws_file_writer_init_internal(writer, allocator, NULL, &s_stdout_writer_vtable);
+}
+
+int aws_stderr_log_writer_init(struct aws_log_writer *writer, struct aws_allocator *allocator) {
+    return s_aws_file_writer_init_internal(writer, allocator, NULL, &s_stderr_writer_vtable);
+}
+
+int aws_file_log_writer_init(
+        struct aws_log_writer *writer,
+        struct aws_allocator *allocator,
+        const char *file_name) {
+    return s_aws_file_writer_init_internal(writer, allocator, file_name, &s_file_writer_vtable);
+}
+
+int aws_log_writer_cleanup(struct aws_log_writer *writer) {
+    return (*writer->vtable->cleanup_fn)(writer);
```

Comment:
Probably want to assert(writer->vtable->cleanup_fn != NULL) here

Reply:
What should be the criteria for determining when/if we check vtable pointers for non-null?  always?

Reply:
Usually on the surface level API functions, e.g. functions called by the user, as high up as we can.